### ISSUE TITLE
Sync TLS certificates

### ISSUE THREAD

**Issue by tam7t** [GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp#96]
### TL;DR

The `kubernetes.io/tls` K8s Secret type expects 2 files `tls.crt` and `tls.key`.

Currently the plugin only allows 1:1 mapping of SM secrets to files. So in order to sync a tls cert for use by an ingress controller, one would have to store the public cert as a separate secret from the private key.

Using 2 secrets and keeping them in sync can be cumbersome.

### Design

**Proposal**

Expand the [provider configuration](https://github.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp/blob/f6264f9af1b9ea3fbed9b8e704a986ba70436d5a/config/config.go#L35-L44) to add a `type` and `field` parameters.

To allow a secret to be interpreted as having a specific format:

```
...
  provider: gcp
  parameters:
    secrets: |
      - resourceName: "projects/$PROJECT_ID/secrets/testsecret/versions/latest"
        fileName: "tls.crt"
        type: kubernetes.io/tls
        field: tls.crt
      - resourceName: "projects/$PROJECT_ID/secrets/testsecret/versions/latest"
        fileName: "tls.key"
        type: kubernetes.io/tls
        field: tls.key
```

Here `projects/$PROJECT_ID/secrets/testsecret/versions/latest` would hold  PEM encoded certificates followed by a PEM encoded key. After accessing the secret the provider would see that it is annotated as type `kubernetes.io/tls` so it would then attempt to parse the certs from the keys and make available `field`.

**Alternatives considered**

The above proposal is still 1:1 secret to files, we could also add some sort of fileMap field:

```
...
  provider: gcp
  parameters:
    secrets: |
      - resourceName: "projects/$PROJECT_ID/secrets/testsecret/versions/latest"
        type: kubernetes.io/tls
        fileMap:
           tls.crt: foo/bar/mycertfile.crt
           tls.key: foo/bar/mykeyfile.crt
```

**Resources**

* https://kubernetes.io/docs/concepts/configuration/secret/
* https://azure.github.io/secrets-store-csi-driver-provider-azure/configurations/ingress-tls/

**Additional information**

**Comment by tam7t** [comment]
I think this actually currently works if the secret holds the key and cert both PEM encoded.

The secret sync feature can pluck key/cert from a single file and split it to the tls type secrets:

```
apiVersion: secrets-store.csi.x-k8s.io/v1alpha1
kind: SecretProviderClass
metadata:
  name: my-tls
spec:
  provider: gcp
  secretObjects:
  - secretName: my-tls-csi
    type: kubernetes.io/tls
    data: 
    - objectName: "mycert.pem"
      key: tls.key
    - objectName: "mycert.pem"
      key: tls.crt
  parameters:
    secrets: |
      - resourceName: "projects/$PROJECT_ID/secrets/testsecret/versions/latest"
         fileName: "mycert.pem"
```

**Comment by tam7t** [comment]
Ok I've confirmed that https://github.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp/issues/96#issuecomment-902969622 works.

The secret value should be the TLS cert + key concatenated and in the PEM format, and the driver will split it when doing secret syncing

**Comment by akarsh** [comment]
@tam7t Please can you elaborate on how the https://github.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp/issues/96#issuecomment-902969622 is working?

**Context**
Currently i'm trying to understand as I had spent two days trying to debug how to get TLS secret from GCP Secret Manager to Ingress Controller using the secrets-store-csi-driver. I could not find any guide related to TLS in the `secrets-store-csi-driver-provider-gcp` GitHub repository.

It would be great if you can answer my questions,

1. Is the mycert.pem file is able to split to tls.cert and tls.key even if the mycert.pem does not contain key value pairs?
for example, mycert.pem is in the following format,
```
-----BEGIN KEY----- 
MIIC2DCCAcCgAwIBAgIBATANBgkqh ...
-----END KEY-----
-----BEGIN CERTIFICATE----- 
MIIC2DCCAcCgAwIBAgIBATANBgkqh ...
-----END CERTIFICATE----
```

sample reference: https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets

2. How are you volume mounting `my-tls` in a pod or deployment yaml file of a helm chart?

3. Do you use `nodePublishSecretRef` to set the secret in the GCP ingress controller TLS secret?

**Comment by akarsh** [comment]
**Update**
I was able to make it work.

> 1. Is the mycert.pem file is able to split to tls.cert and tls.key even if the mycert.pem does not contain key value pairs?
for example, mycert.pem is in the following format,

Yes the mycert.pem file is able to split into  tls.cert and tls.key. You need to make sure the mycert.pem file is in the following format,

```
-----BEGIN KEY----- 
MIIC2DCCAcCgAwIBAgIBATANBgkqh ...
-----END KEY-----
-----BEGIN CERTIFICATE----- 
MIIC2DCCAcCgAwIBAgIBATANBgkqh ...
-----END CERTIFICATE----
```

> 2. How are you volume mounting my-tls in a pod or deployment yaml file of a helm chart?

You need to volume mount the mycert.pem file which is similar to mounting secrets as mentioned in the example [examples/mypod.yaml.tmpl](https://github.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp/blob/main/examples/mypod.yaml.tmpl).

> 3. Do you use nodePublishSecretRef to set the secret in the GCP ingress controller TLS secret?

Not required, because the mycert.pem fille will be added as a tls secret in the kubernetes cluster. This is done by the following code from the my-tls secret provider class,
```
secretObjects:
  - secretName: my-tls-csi
    type: kubernetes.io/tls
    data: 
    - objectName: "mycert.pem"
      key: tls.key
    - objectName: "mycert.pem"
      key: tls.crt
```

Note: Please make sure the new files are mounted in a different folder then the existing secrets if you are performing helm upgrade if not then you can mount all files in a single folder or other folder whichever works for you.
